### PR TITLE
refactor: Migrate ConversationRepository to TypeScript

### PR DESCRIPTION
**Pull request by bennyn**

### CODE REVIEW

**Review by ffflorian — APPROVED** [review]

File: src/script/entity/message/Message.ts [hunk 1]
```diff
@@ -70,6 +77,7 @@ export class Message {
   public readReceipts: ko.ObservableArray<ReadReceipt>;
   public super_type: SuperType;
   public type: string;
+  reaction: any;
```
Comment (thisisamir98) [inline src/script/entity/message/Message.ts]: can't we have a type for this?

Comment (ffflorian) [reply src/script/entity/message/Message.ts]: Yeah, this is probably `ReactionType` (from `@wireapp/core/dist/conversation`).

File: src/script/view_model/content/MessageListViewModel.ts [hunk 2]
```diff
@@ -330,7 +330,7 @@ export class MessageListViewModel {
     if (lastMessage) {
       if (!this.isLastReceivedMessage(lastMessage, this.conversation())) {
         // if the last loaded message is not the last of the conversation, we load the subsequent messages
-        return this.conversationRepository.getSubsequentMessages(this.conversation(), lastMessage, false);
```
Comment (thisisamir98) [inline src/script/view_model/content/MessageListViewModel.ts:333]: @AndyLnd are you sure removing this return statement is safe?

File: src/script/conversation/ConversationRepository.ts [hunk 3]
```diff
@@ -1063,7 +1092,7 @@ export class ConversationRepository {
    * @param {string} message_id Message ID
    * @returns {Promise} Resolves with `true` if message is marked as read
    */
-  isMessageRead(conversation_id, message_id) {
+  isMessageRead(conversation_id: any, message_id: string) {
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
  isMessageRead(conversation_id: string, message_id: string) {
```

File: src/script/conversation/ConversationRepository.ts [hunk 4]
```diff
@@ -1168,11 +1198,10 @@ export class ConversationRepository {
   /**
    * Maps user connections to the corresponding conversations.
    * @param {Array<ConnectionEntity>} connectionEntities Connections entities
-   * @returns {undefined} No return value
    */
-  map_connections(connectionEntities) {
+  map_connections(connectionEntities: any[]) {
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
  map_connections(connectionEntities: ConnectionEntity[]) {
```

File: src/script/conversation/ConversationRepository.ts [hunk 5]
```diff
@@ -1336,7 +1366,7 @@ export class ConversationRepository {
       .catch(error => this._handleAddToConversationError(error, conversationEntity, userIds));
   }
 
-  addMissingMember(conversationEntity, userIds, timestamp) {
+  addMissingMember(conversationEntity: Conversation, userIds: any[], timestamp: number) {
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
  addMissingMember(conversationEntity: Conversation, userIds: string[], timestamp: number) {
```

File: src/script/conversation/ConversationRepository.ts [hunk 6]
```diff
@@ -1168,11 +1198,10 @@ export class ConversationRepository {
   /**
    * Maps user connections to the corresponding conversations.
    * @param {Array<ConnectionEntity>} connectionEntities Connections entities
-   * @returns {undefined} No return value
    */
-  map_connections(connectionEntities) {
+  map_connections(connectionEntities: any[]) {
     this.logger.info(`Mapping '${connectionEntities.length}' user connection(s) to conversations`, connectionEntities);
-    connectionEntities.map(connectionEntity => this.map_connection(connectionEntity));
+    connectionEntities.map((connectionEntity: any) => this.map_connection(connectionEntity));
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
    connectionEntities.map((connectionEntity: ConnectionEntity) => this.map_connection(connectionEntity));
```

File: src/script/conversation/ConversationRepository.ts [hunk 7]
```diff
@@ -4121,21 +4191,19 @@ export class ConversationRepository {
   /**
    * Cancel asset upload.
    * @param {string} messageId Id of the message which upload has been cancelled
-   * @returns {undefined} No return value
    */
-  cancel_asset_upload(messageId) {
+  cancel_asset_upload(messageId: any) {
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
  cancel_asset_upload(messageId: string) {
```

File: src/script/conversation/ConversationRepository.ts [hunk 8]
```diff
@@ -1980,7 +2008,11 @@ export class ConversationRepository {
    * @param {ProtobufAsset.NotUploaded} [reason=ProtobufAsset.NotUploaded.FAILED] Cause for the failed upload (optional)
    * @returns {Promise} Resolves when the asset failure was sent
    */
-  send_asset_upload_failed(conversationEntity, messageId, reason = ProtobufAsset.NotUploaded.FAILED) {
+  send_asset_upload_failed(
+    conversationEntity: Conversation,
+    messageId: any,
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
    messageId: string,
```

File: src/script/conversation/ConversationRepository.ts [hunk 9]
```diff
@@ -2164,7 +2203,7 @@ export class ConversationRepository {
    * @param {number} zoom Zoom factor for the map (Google Maps)
    * @returns {Promise} Resolves after sending the location
    */
-  sendLocation(conversationEntity, longitude, latitude, name, zoom) {
+  sendLocation(conversationEntity: Conversation, longitude: any, latitude: any, name: any, zoom: any) {
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
  sendLocation(conversationEntity: Conversation, longitude: number, latitude: number, name: string, zoom: number) {
```

File: src/script/conversation/ConversationRepository.ts [hunk 10]
```diff
@@ -2354,14 +2406,14 @@ export class ConversationRepository {
...
-    quoteEntity,
-    linkPreviews,
-    expectsReadConfirmation,
-    legalHoldStatus,
+  private _createTextProto(
+    messageId: string,
+    textMessage: string,
+    mentionEntities: MentionEntity[],
+    quoteEntity: QuoteEntity,
+    linkPreviews: any[],
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]: type for this should be LinkPreview[] and can be imported from `LinkPreviewRepository.ts`

File: src/script/conversation/ConversationRepository.ts [hunk 11]
```diff
@@ -2354,14 +2406,14 @@ export class ConversationRepository {
...
+    expectsReadConfirmation: boolean,
+    legalHoldStatus: any,
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]: type for this is `LegalHoldStatus` and can be imported from `@wireapp/protocol-messaging`

File: src/script/conversation/ConversationRepository.ts [hunk 12]
```diff
@@ -3172,18 +3242,21 @@ export class ConversationRepository {
 
         return conversationEntity;
       })
-      .then(conversationEntity => this._checkLegalHoldStatus(conversationEntity, eventJson))
-      .then(conversationEntity => this._checkConversationParticipants(conversationEntity, eventJson, eventSource))
-      .then(conversationEntity => this._triggerFeatureEventHandlers(conversationEntity, eventJson, eventSource))
-      .then(conversationEntity => this._reactToConversationEvent(conversationEntity, eventJson, eventSource))
+      .then((conversationEntity: any) => this._checkLegalHoldStatus(conversationEntity, eventJson))
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
      .then((conversationEntity: Conversation) => this._checkLegalHoldStatus(conversationEntity, eventJson))
```

File: src/script/conversation/ConversationRepository.ts [hunk 13]
```diff
@@ -3172,18 +3242,21 @@ export class ConversationRepository {
...
+      .then((conversationEntity: any) =>
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
      .then((conversationEntity: Conversation) =>
```

File: src/script/conversation/ConversationRepository.ts [hunk 14]
```diff
@@ -3172,18 +3242,21 @@ export class ConversationRepository {
...
+        this._checkConversationParticipants(conversationEntity, eventJson, eventSource),
+      )
+      .then((conversationEntity: any) => this._triggerFeatureEventHandlers(conversationEntity, eventJson))
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
      .then((conversationEntity: Conversation) => this._triggerFeatureEventHandlers(conversationEntity, eventJson))
```

File: src/script/conversation/ConversationRepository.ts [hunk 15]
```diff
@@ -3172,18 +3242,21 @@ export class ConversationRepository {
...
+      .then((conversationEntity: any) => this._reactToConversationEvent(conversationEntity, eventJson, eventSource))
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
      .then((conversationEntity: Conversation) => this._reactToConversationEvent(conversationEntity, eventJson, eventSource))
```

File: src/script/conversation/ConversationRepository.ts [hunk 16]
```diff
@@ -3294,13 +3370,12 @@ export class ConversationRepository {
...
    * Triggers the methods associated with a specific event.
    *
-   * @private
    * @param {Conversation} conversationEntity Conversation targeted by the event
    * @param {Object} eventJson JSON data of the event
    * @param {EventRepository.SOURCE} eventSource Source of event
    * @returns {Promise<any>} Resolves when the event has been treated
    */
-  _reactToConversationEvent(conversationEntity, eventJson, eventSource) {
+  private _reactToConversationEvent(conversationEntity: any, eventJson: EventJson, eventSource: EventSource) {
```
Comment (thisisamir98) [inline src/script/conversation/ConversationRepository.ts]:
```suggestion
  private _reactToConversationEvent(conversationEntity: Conversation, eventJson: EventJson, eventSource: EventSource) {
```

File: src/script/view_model/content/MessageListViewModel.ts (hunk 2 above)
Comment (ffflorian) [reply src/script/view_model/content/MessageListViewModel.ts:333]: I used `else if` instead.

**Pull request merged**